sogs/events.py

```diff
@@ -0,0 +1,160 @@
+from collections import defaultdict
+
+from oxenmq import AuthLevel
+
+from . import config
+from . import model
+from .web import app
+from .omq import omq
+
+from binascii import hexlify
+import json
+import sys
+
+# pools for event propagation
+_pools = defaultdict(list)
+
+status_OK = 'OK'
+status_ERR = 'ERROR'
+
+# the events we are able to subscribe to
+EVENTS = ('message', 'joined', 'parted', 'banned', 'unbanned', 'deleted', 'uploaded')
+
+
+def event_name_valid(eventname):
+    """ return True if this event name is something well formed """
+    return eventname.lower() in EVENTS
+
+
+def _user_from_conn(conn, automod=True):
+    """
+    make a model.User from a connection using it's curve pubkey as the session id, if automod is
+    True we auto promote the bot to a moderator if it isn't already
+    """
+    user = model.User(session_id='05' + hexlify(conn.pubkey).decode())
+    if automod and not user.global_moderator:
+        user.set_moderator()
+    return user
+
+
+def _sub_conn_to_event(name, conn):
+    """ subscribe a connection to an event type """
+    if not event_name_valid(name):
+        raise Exception('invalid event name: {}'.format(name))
+    global _pools
+    _pools[name].append(conn)
+
+
+def _handle_subscribe_request(msg):
+    """ handle a request to subscribe to events as a bot"""
+    parts = msg.data()
+    if len(parts) == 0:
+        raise Exception("no events selected")
+
+    for name in parts:
+        name = name.decode()
+        if name in ('all', '*'):
+            # subscribe to all events
+            for ev in EVENTS:
+                _sub_conn_to_event(ev, msg.conn)
```

Comment:
I don't think this should be allowed because, if it's used, a future SOGS update could add new events that you don't know how to handle.

Reply:
i agree